### PR TITLE
added grpc

### PR DESCRIPTION
**Pull request by afanasievM**
1. In transaction service added server
2. In university service added client now without tests
3. **without tests**. Will be later if needs

### CODE REVIEW

File: transaction-service/src/main/kotlin/protobuf/GRPCService.proto [hunk 1]
```diff
@@ -0,0 +1,9 @@
+syntax = "proto3";
```
Comment (ajax-shulepa-a) [inline transaction-service/src/main/kotlin/protobuf/GRPCService.proto]: proto files should be placed in separate module, so client which will use your API can add dependency only on that module
for more details about naming messages and organising files search 'Internal API conventions (v2)' on confluence

File: transaction-service/src/main/kotlin/protobuf/GRPCService.proto [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+package protobuf;
+
+import "ProtoMessage.proto";
+
+service TransactionsService{
+  rpc getTransactions (TransactionRequestProto) returns (stream TransactionResponseProto);
```
Comment (ajax-shulepa-a) [inline transaction-service/src/main/kotlin/protobuf/GRPCService.proto]: if your service returns many elements at once and then does nothing - it is not a server-side stream. In such case, it should be a plain request-reply where the reply contains a list of elements
The stream should be used when the server will publish updates to stream in the future

File: university-service/src/main/kotlin/protobuf/GRPCService.proto [hunk 3]
```diff
@@ -0,0 +1,9 @@
+syntax = "proto3";
+
+package protobuf;
+
+import "ProtoMessage.proto";
+
+service TransactionsService{
```
Comment (ajax-shulepa-a) [inline university-service/src/main/kotlin/protobuf/GRPCService.proto]: you shouldn't copy the same proto in each module. proto files should be placed in separate module and other module can add it as dependency

**Review by ajax-shulepa-a — CHANGES_REQUESTED** [review]

File: university-service/src/main/kotlin/protobuf/GRPCService.proto (hunk 3 above)
Comment (afanasievM) [reply university-service/src/main/kotlin/protobuf/GRPCService.proto]: okay, done

File: transaction-service/src/main/kotlin/protobuf/GRPCService.proto (hunk 2 above)
Comment (afanasievM) [reply transaction-service/src/main/kotlin/protobuf/GRPCService.proto]: changed to list

File: transaction-service/src/main/kotlin/protobuf/GRPCService.proto (hunk 1 above)
Comment (afanasievM) [reply transaction-service/src/main/kotlin/protobuf/GRPCService.proto]: renamed

File: transaction-service/build.gradle.kts [hunk 4]
```diff
@@ -105,10 +111,22 @@ kapt {
 
 protobuf {
     protoc {
-        artifact = "com.google.protobuf:protoc:3.19.6"
+        artifact = "com.google.protobuf:protoc:3.24.3"
+    }
+    plugins {
```
Comment (ajax-shulepa-a) [inline transaction-service/build.gradle.kts]: move proto-related tasks to protobuf module

File: transaction-service/proto/Imports.proto [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import "protobuf/GRPCService.proto";
```
Comment (ajax-shulepa-a) [inline transaction-service/proto/Imports.proto]: it's not needed

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/gprc/TransactionGRPCService.kt [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            transactionTimeBuilder
+                .setSeconds(dto.transactionTime?.toEpochSecond(ZoneOffset.UTC) ?: 0)
+                .setNanos(dto.transactionTime?.nano ?: 0)
+            value = dto.value ?: 0.0
+            currency = dto.currency
+            iban = dto.iban
+        }.build()
+    }
+
+}
```
Comment (ajax-shulepa-a) [inline transaction-service/src/main/kotlin/ua/com/foxminded/restClient/gprc/TransactionGRPCService.kt]:
```suggestion
    }
}
```

File: university-service/build.gradle.kts [hunk 7]
```diff
@@ -84,3 +102,28 @@ tasks.withType<Detekt>().configureEach {
 tasks.withType<DetektCreateBaselineTask>().configureEach {
     jvmTarget = "17"
 }
+
+protobuf {
```
Comment (ajax-shulepa-a) [inline university-service/build.gradle.kts]: these tasks should be moved to protobuf module

**Review by ajax-shulepa-a — CHANGES_REQUESTED** [review]

File: transaction-service/proto/Imports.proto (hunk 5 above)
Comment (afanasievM) [reply transaction-service/proto/Imports.proto]: It cant find message without import

File: transaction-service/build.gradle.kts (hunk 4 above)
Comment (afanasievM) [reply transaction-service/build.gradle.kts]: ok

File: university-service/build.gradle.kts (hunk 7 above)
Comment (afanasievM) [reply university-service/build.gradle.kts]: ok

File: transaction-service/proto/Imports.proto (hunk 5 above)
Comment (afanasievM) [reply transaction-service/proto/Imports.proto]: done

File: protobuf/.gitattributes [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#
```
Comment (ajax-shulepa-a) [inline protobuf/.gitattributes]: this file should be present only in root module

File: protobuf/.gitignore [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Ignore Gradle project-specific cache directory
+.gradle
```
Comment (ajax-shulepa-a) [inline protobuf/.gitignore]: this file should be present only in root module

File: protobuf/build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,69 @@
+import com.google.protobuf.gradle.id
+import com.google.protobuf.gradle.proto
+
+
+
+
```
Comment (ajax-shulepa-a) [inline protobuf/build.gradle.kts]: too many empty lines

File: protobuf/build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+sourceSets {
+    main {
+        proto {
+            srcDir("/proto")
+        }
+    }
+}
+
+group = "ua.com.foxminded"
+version = "0.0.1-SNAPSHOT"
```
Comment (ajax-shulepa-a) [inline protobuf/build.gradle.kts]: group and version should be defined in root gradle and reused in submodules

File: protobuf/gradlew [hunk 12]
```diff
@@ -0,0 +1,234 @@
+#!/bin/sh
```
Comment (ajax-shulepa-a) [inline protobuf/gradlew]: gradlew should be only in root module

File: university-service/src/main/kotlin/ua/com/foxminded/courseproject/service/TransactionGRPCServiceImpl.kt [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+                ZoneOffset.UTC
+            )
+            transactionDirection = Direction.valueOf(responseProto.direction.toString())
+            value = responseProto.value
+            currency = responseProto.currency
+            iban = responseProto.iban
+        }
+    }
+
+}
```
Comment (ajax-shulepa-a) [inline university-service/src/main/kotlin/ua/com/foxminded/courseproject/service/TransactionGRPCServiceImpl.kt]:
```suggestion
    }
}
```
before closing brace should be no empty line

**Review by ajax-shulepa-a — COMMENTED** [review]
Good job, let's just fix some technical issues

File: protobuf/build.gradle.kts (hunk 11 above)
Comment (afanasievM) [reply protobuf/build.gradle.kts]: ок

File: protobuf/gradlew (hunk 12 above)
Comment (afanasievM) [reply protobuf/gradlew]: ок

File: protobuf/build.gradle.kts (hunk 10 above)
Comment (afanasievM) [reply protobuf/build.gradle.kts]: fixed

File: protobuf/.gitignore (hunk 9 above)
Comment (afanasievM) [reply protobuf/.gitignore]: ok

File: protobuf/.gitattributes (hunk 8 above)
Comment (afanasievM) [reply protobuf/.gitattributes]: ok

**Review by ajax-shulepa-a — APPROVED** [review]

**Pull request merged**